cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -434,14 +434,15 @@ namespace {
 
 Result<std::shared_ptr<io::OutputStream>> OpenOutputStreamGeneric(const std::string& path,
                                                                   bool truncate,
-                                                                  bool append) {
+                                                                  bool append,
+                                                                  bool reuse) {
   RETURN_NOT_OK(ValidatePath(path));
   ARROW_ASSIGN_OR_RAISE(auto fn, PlatformFilename::FromString(path));
   const bool write_only = true;
-  ARROW_ASSIGN_OR_RAISE(
-      auto fd, ::arrow::internal::FileOpenWritable(fn, write_only, truncate, append));
+  ARROW_ASSIGN_OR_RAISE(auto fd, ::arrow::internal::FileOpenWritable(
+                                     fn, write_only, truncate, append, !reuse));
```

Comment:
```suggestion
  // When reuse is false we need to open the file in sync mode because fadvise will
  // not drop the data from cache unless it has finished writing.
  ARROW_ASSIGN_OR_RAISE(auto fd, ::arrow::internal::FileOpenWritable(
                                     fn, write_only, truncate, append, !reuse));
```